### PR TITLE
Feature/quadtrees

### PR DESCRIPTION
**Pull request by PearCoding**
As the name implies. Support for DQT

### CODE REVIEW

File: OpenPGL.NET/Field.cs [hunk 1]
```diff
@@ -200,15 +210,51 @@ public class VMMDirectionalSettings : DirectionalSettings {
...
+        Marshal.StructureToPtr<OpenPGL.PGLDQTFactoryArguments>(args, target.directionalDistributionArguments, true);
+    }
+}
+
 public struct FieldSettings {
     public SpatialSettings SpatialSettings { get; init; }
     public DirectionalSettings DirectionalSettings { get; init; }
 
     internal OpenPGL.PGLFieldArguments MakeArguments() {
+        // We require the correct struct and dist type for preallocation done inside pglFieldArgumentsSetDefaults
```
Comment (pgrit) [inline OpenPGL.NET/Field.cs]:
```suggestion
       
```

File: OpenPGL.NET/Field.cs [hunk 2]
```diff
@@ -200,15 +210,51 @@ public class VMMDirectionalSettings : DirectionalSettings {
...
         OpenPGL.PGLFieldArguments arguments;
         OpenPGL.pglFieldArgumentsSetDefaults(out arguments,
-            OpenPGL.PGL_SPATIAL_STRUCTURE_TYPE.PGL_SPATIAL_STRUCTURE_KDTREE,
-            OpenPGL.PGL_DIRECTIONAL_DISTRIBUTION_TYPE.PGL_DIRECTIONAL_DISTRIBUTION_PARALLAX_AWARE_VMM);
+            SpatialSettings.GetStructType(),
```
Comment (pgrit) [inline OpenPGL.NET/Field.cs]: If you want to do that, the properties above need a non-null default. Else this will crash if the user tries to run with default settings.

File: OpenPGL.NET/Field.cs [hunk 3]
```diff
@@ -114,10 +114,12 @@ public struct BoundingBox {
 }
 
 public abstract class SpatialSettings {
+    internal abstract OpenPGL.PGL_SPATIAL_STRUCTURE_TYPE GetStructType();
```
Comment (pgrit) [inline OpenPGL.NET/Field.cs]:
```suggestion
    internal abstract OpenPGL.PGL_SPATIAL_STRUCTURE_TYPE StructType { get; }

```

File: OpenPGL.NET/Field.cs [hunk 4]
```diff
@@ -114,10 +114,12 @@ public struct BoundingBox {
 }
 
 public abstract class SpatialSettings {
+    internal abstract OpenPGL.PGL_SPATIAL_STRUCTURE_TYPE GetStructType();
     internal abstract void SetArguments(ref OpenPGL.PGLFieldArguments target);
 }
 
 public abstract class DirectionalSettings {
+    internal abstract OpenPGL.PGL_DIRECTIONAL_DISTRIBUTION_TYPE GetDistType();
```
Comment (pgrit) [inline OpenPGL.NET/Field.cs]:
```suggestion
    internal abstract OpenPGL.PGL_DIRECTIONAL_DISTRIBUTION_TYPE DistType { get; }

```

**Review by pgrit — CHANGES_REQUESTED** [review]

File: OpenPGL.NET/Field.cs (hunk 2 above)
Comment (PearCoding) [reply OpenPGL.NET/Field.cs]: Ohh, you're right. Will fix that

File: OpenPGL.NET/Field.cs (hunk 1 above)
Comment (PearCoding) [reply OpenPGL.NET/Field.cs]: I would rather keep the comment (maybe rephrase it). As it is not really obvious that `pglFieldArgumentsSetDefaults` uses new allocation to setup the structs. It is also something a common C API would NOT do. (as far as my experience goes)

Comment (pgrit) [reply OpenPGL.NET/Field.cs]: Since the memory is freed by device.NewField, probably best to move this logic to the constructor of field in the wrapper like
```csharp
public Field(FieldSettings settings = new()) {
        device = new();
        
        // pglFieldArgumentsSetDefaults allocates memory for the settings, that we can then write to.
        // pglDeviceNewField frees that memory again
        OpenPGL.pglFieldArgumentsSetDefaults(out var arguments, 
            settings.SpatialSettings.StructType, settings.DirectionalSettings.DistType);
        SpatialSettings.SetArguments(ref arguments);
        DirectionalSettings.SetArguments(ref arguments);

        Handle = OpenPGL.pglDeviceNewField(device.Ptr, arguments);
        Debug.Assert(Handle != nint.Zero);
    }

```

Comment (PearCoding) [reply OpenPGL.NET/Field.cs]: Yeah, thats a better approach

**Review by pgrit — APPROVED** [review]
